### PR TITLE
HIVE-25235: Remove ThreadPoolExecutorWithOomHook

### PR DESCRIPTION
**Pull request by belugabehr**

### CODE REVIEW

**Comment by belugabehr** [comment]
@nrg4878 @dengzhhu653 Care to take a peek?

**Comment by dengzhhu653** [comment]
The oom runner would make the hs2 shutdown gracefully after releasing resources, such as the leadership of hs2,  some sessions related, the instances with oom may not be able to exit without the appropriate jvm configs, but it's ok to me that adding the `-XX:+ExitOnOutOfMemoryError`(or better make it default?) to avoid the long running oom instances, just my opinion, hope to hear from others.
Thanks,
Zhihua Deng

**Comment by belugabehr** [comment]
@dengzhhu653 Hey, thanks for the feedback!

My issue here is that, the best way to try to trigger a graceful shutdown process is with the `ExitOnOutOfMemoryError` approach.  This OOM feature gets triggered by the JVM itself when the application's heap memory is exhausted.  The problem with the current approach is that it's quite a bit of code which runs in an indeterminate way.  If the application is out of memory, it is out of memory.  Doing things like gracefully shutting down may themselves require a small amount of memory to complete and therefore are at risk of failing.  So, ultimately it's a bunch of code that may or may not actually work when its needed. Besides, things like ZooKeeper are very adapt at handling servers crashing.  Yes, it's "nice" to close gracefully, but it's not really needed.  It's probably better to just remove the OOM-ing instance of HS2 out of the cluster as fast as possible, instead of trying to limp along and shutdown nicely.

Also, this code is only helpful for threads from these pools, but what if things fails from threads from other pools?  What if the OOM comes from Jetty server itself (and not one of the worker threads)?  Again, it's best for the JVM to handle this stuff since it is pretty immune from an exhausted heap. A bunch of code here with questionable usefulness.

**Comment by dengzhhu653** [comment]
> Doing things like gracefully shutting down may themselves require a small amount of memory to complete and therefore are at risk of failing
Yes, there is a chance that the shutting down does not execute completely, waiting for next oom thrown from the hs2's handler pools may be unreliable, there are some shutdown hooks registered on startup of hs2, this may take over gracefully shutting down when hs2 is about to exit.

File: service/src/java/org/apache/hive/service/cli/session/SessionManager.java [hunk 1]
```diff
@@ -224,7 +224,7 @@ private void createBackgroundOperationPool() {
     // Threads terminate when they are idle for more than the keepAliveTime
     // A bounded blocking queue is used to queue incoming operations, if #operations > poolSize
     String threadPoolName = "HiveServer2-Background-Pool";
-    final BlockingQueue queue = new LinkedBlockingQueue<Runnable>(poolQueueSize);
+    final BlockingQueue<Runnable> queue = new LinkedBlockingQueue<Runnable>(poolQueueSize);
```
Comment (miklosgergely) [inline service/src/java/org/apache/hive/service/cli/session/SessionManager.java:227]: You can use <> here

**Review by miklosgergely — APPROVED** [review]

File: service/src/java/org/apache/hive/service/cli/session/SessionManager.java (hunk 1 above)
Comment (belugabehr) [reply service/src/java/org/apache/hive/service/cli/session/SessionManager.java:227]: I'm going to commit these changes without this one change to save us from having to re-build :)

**Pull request merged**